### DISCUSSION TITLE
[Kavita] Highlighting / saving selected text / annotating for epub and pdf

### DISCUSSION THREAD

**Discussion by DieselTech** [Kareadita/Kavita#2543] [Ideas]
### Idea Description

Posted by Starry_air
· Mar 2022
· 23 Upvotes
https://feats.kavitareader.com/posts/86/highlighting-saving-selected-text-annotating-for-epub-and-pdf

Description

As the server doesn't perform any writing operations on the files, saving highlights / annotations cannot be done in the epub/pdf files themselves, but perhaps they could be stored in the database with a browse/export function? Or perhaps, with time, even in the files themselves?

### Duration of Using Kavita

42

### Idea Category

Feature Enhancement

### Agreement

- [X] I agree that this is solely for submitting ideas, and I will search for existing ideas before posting.

**Comment by ProfDoof** [comment]
This may just be me, but ideally I would prefer if the annotations are alongside or in the file rather than in the database as I would like to be able to just grab the file/folder and be able to put that anywhere and it work. 

As for the annotations themselves, ideally they would allow for
- Various highlight colors (maybe even a color picker)
- Me to see the highlighted text and the comment that was added
- Me to associate a concept or context with highlight colors and create an annotation legend (just something I've missed in the switch from paper to digital)
- Underline and not just highlight
- Link annotations together directly rather than just via color

My MVP would be
- Highlight with some distinct neon color
- Associate a comment with a highlight

**Comment by Menghini** [comment]
I second the need for this. Moreover, I think it would beneficial to be a database or something similar because you will then be able have more than one user add highlights and annotate/comment in the book.

**Comment by AlbertMarkovski** [comment]
Thirding the need for this. Database would be ideal so as to leave original ebook files unmodified (I prefer to mount them over NFS as read only). 

Seems like lots of people are looking for such a thing: 

https://old.reddit.com/r/selfhosted/comments/12gybq2/looking_for_a_pdfepub_reader_with_annotation_and/ 

I gave Kavita a test drive anyway and it's slick otherwise!

**Comment by jvarghe** [comment]
Fourthing this. I came in from the link posted by AlbertMarkovski. It lays out everything that's missing in ebook readers nicely. 

I want to be able to highlight (in various colours) bookmark and annotate all kinds of books (pdfs, epubs etc.). 

It would be good to leave the original files untouched, and yet have annotations that are portable, but these requirements might conflict. 

I don't want to leave my annotations trapped in one any app. Has anyone looked into the [Web Annotation](https://en.wikipedia.org/wiki/Web_annotation) standard? Perhaps implementing that would make it easier to export annotations.

Even more advanced features, like searching your annotations database, and collating and linking annotations across books would be great.

**Comment by aOk88** [comment]
+1 for this. 

Being able to see "highlights" on your bookmarks would be ideal within the book. Apple's Books does this (see attached). Another thing that would be useful is if in the menu where it shows "Special", "Reviews" and "Details" it can also show bookmarks too.

![IMG_0278](https://github.com/user-attachments/assets/4196b15a-6c85-4da2-a1ad-01e1e66fb4f7)

**Comment by fvelo** [comment]
I personally think that the ability to highlight is super useful. I understand that, from what I have seen, Kavita is designed primarily for comics, but if someone uses it to study books on the go it is a must.

I don't know if it is possible or wise, but perhaps saving highlights in the book metadata might be a possibility to ensure they are different for each user and for easy portability?

**Comment by majora2007** [comment]
For people that use highlighting/annotations, what is the primary use case and interactions around the feature? 

How might you envision it within Kavita? 

We already have a bookmark system in Kavita for books, highlighting has always been something I was interested in tackling. But I need a better understanding before I take a stab at implementing.

**Reply by Doomwhite** [reply]
How it works in calibre is a good approach IMO, there you can make multiple custom colors options, and add a comment to the highlight. Then it's available in a table of highlights where you can check the book of the highlight, the type(color), and the comment, if there's one.

Creating a highlight:
![image](https://github.com/user-attachments/assets/0b3c02ee-1f4a-4841-8f5b-5ee502c2fc9c)

Viewing the highlight with a comment:
![image](https://github.com/user-attachments/assets/c1f8f1f9-f0f6-4a44-83f2-dbdee77b63ab)

Viewing the "table" of highlights/annotations:
![image](https://github.com/user-attachments/assets/3d53f653-a108-49ac-b078-c345449139cb)

A way to export/import those annotations is very powerful, but also annoying to implement (I would imagine EPUB is a bit more annoying to sync to the actual view.)?

In calibre I can export those notes to obsidian, like this, but that could be solved with a template-like export:
![image](https://github.com/user-attachments/assets/8e67c42c-70aa-4ee8-b583-3115007d67ad)

**Reply by majora2007** [reply]
This is very handy. 

When you are looking at these annotations/highlights, I take it you are in a "mode" rather than a text mode. I was thinking of providing some icon to indicate to the user that there is a highlight/annotation here and clicking it (or switching into an annotation mode) would render the highlighting color and/or annotations. 

For the export that you do, what's the purpose that it solves?

**Reply by Doomwhite** [reply]
> When you are looking at these annotations/highlights, I take it you are in a "mode" rather than a text mode. I was thinking of providing some icon to indicate to the user that there is a highlight/annotation here and clicking it (or switching into an annotation mode) would render the highlighting color and/or annotations.

The highlights are always visible, but having a toggleable mode would be a better interface, I guess.

> For the export that you do, what's the purpose that it solves?

In my case, it's useful for the purpose of having the notes in another UI.
It basically exports to a md format, from there you can pretty much use it in any website/annotations software, notice how calibre has an URL to the highlight in the book, in the case of Kavita, I don't know if that level of detail is needed.

**Reply by majora2007** [reply]
Coming back to looking at Calibre, how often would you say the advanced formatting is used? Like strikethru, etc? Or is text more than enough? 

Would you forsee a benefit of seeing other users on your server's annotations (assuming they opt in to sharing and you opt-in to seeing)?

**Reply by scare376** [reply]
> Would you forsee a benefit of seeing other users on your server's annotations (assuming they opt in to sharing and you opt-in to seeing)?

Absolutely. i think we discussed at some point including features for like book clubs and such. being able to share annotations would be huge for discussions. Also personally for my use case my library is geared towards japanese learning so seeing someone else's annotations would be very helpful

**Reply by scare376** [reply]
one thing i thought of if we do sharing. we should probably have a way to spoiler tag it. (like how discord does it where you cant see the text unless you click on it)

**Reply by Doomwhite** [reply]
> Coming back to looking at Calibre, how often would you say the advanced formatting is used? Like strikethru, etc? Or is text more than enough?

IMO underline/strikethrough is pretty useful.

**Comment by majora2007** [comment]
FYI, I'm looking to get this into v0.8.8. I did some quick proof of concept work and my solution works. 

This is the time to provide me with as much consideration as possible before I implement the feature (as the usual circle around to rework features is 3 years).

**Reply by 2302Dave** [reply]
Do you have any updates on this topic? Will it be released with 0.8.8?

**Reply by majora2007** [reply]
It is currently in development and will likely be released with v0.8.8. It depends on how user feedback goes.

**Reply by 2302Dave** [reply]
Is still feedback needed? When will v0.8.8 be released?

**Reply by majora2007** [reply]
Released when it's ready. I haven't launched the first pass yet. See the comment below about how to join for feedback.

**Reply by scare376** [reply]
we dont provide timeframes on releases. it will be released when it is ready. the feature will be pushed into a nightly build soon to get feedback from nightly users and clean up bugs before going to stable

**Comment by majora2007** [comment]
If anyone here is in our discord, please dm me. I would like to work more closely with you to help build this epub highlight/annotation system.

**Comment by majora2007** [comment]
A nightly was released with the annotation system and I am accepting feedback in the Annotation [discord channel](https://discord.gg/T8Ux6FMKVc). Please note that the reader is pretty shaky due to the MASSIVE amount of refactoring that took place.

**Reply by majora2007** [reply]
Update: We have stabilized and the annotation system is nearly complete. We have the ability to filter in the UI with custom filtering/sorting and exporting (filtered or everything). 

We have one social feature planned (opt-in) for Annotations then this will be considered complete. Thanks for all the feedback.

**Comment by majora2007** [comment]
We have switched to code freeze, this will be delivered in v0.8.8. 

While PDF is listed here, it is unfortunately not possible at this time due to the library we use. Feel free to open a new FR with Annotations for PDF.
